FEAT: Upgrade to .NET 10

matthewtoghill opened the pull request:
## Description
Upgrade the solution to .NET 10.

- chore: Packages updated
- fix: Breaking changes resolved with:
  - MockQueryable - API change related to `AsQueryable()`
  - Swashbuckle & Microsoft.OpenApi - API changes related to upgrading the OpenApi version
- fix: remove `[FromFile]` annotation on endpoint `IFormFile` parameter
- feat: make use of new unbound generic type support for `nameof`
- feat: move common MSBuild properties to Directory.Build.props file

## Related Issue
Closes #101 

## How Has This Been Tested?
Tests pass
Test apps created from template created successfully, build and run as expected

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

matthewtoghill committed:
feat: update target framework on projects to dotnet 10

matthewtoghill committed:
feat: update release pipeline to use dotnet-version 10.x

matthewtoghill committed:
chore: update readme

matthewtoghill committed:
fix: remove FromForm annotation on endpoint for IFormFile parameter

matthewtoghill committed:
chore: update packages and resolve breaking changes

matthewtoghill committed:
feat: update use of nameof with support for unbound generic types

matthewtoghill committed:
chore: formatting

CesarD committed:
Applying C# 14 new features to code, such as extension blocks and other small refactors to update the current code with the most recent language features.

CesarD committed:
Resolving merge conflicts.

CesarD committed:
Updated dependencies.
Added CompatibilityLevel configuration to EF Core configuration for most recent features usages.

matthewtoghill committed:
chore: move common MSBuild properties to Directory.Build.props file

matthewtoghill committed:
chore: update packages

matthewtoghill commented on src/Content/Backend/Solution/Monaco.Template.Backend.sln:
Now that we've upgraded to .NET10 in this PR, we need to decide if we want to continue to keep a `.sln` file for compatibility or if it's possible to remove it and just have the new `.slnx` file.

CesarD replied:
Yup, I believe we can get rid of the .sln file for good 🙂

matthewtoghill replied:
I've removed the .sln file but I think the templating has not fixed the issue with XML conditional processing in .slnx files. We have a solution to that though so it is not a problem.

CesarD replied:
Sorry, what problem was it?

matthewtoghill replied:
Within the `.slnx` file we have the `<!--#if -->` blocks. These were not being processed in the expected way for this file type when we first added the `.slnx` file. We have a `SpecialCustomOperations` item in the `template.json` file that deals with handling these blocks in the `.slnx` file.

I've not tested if the templating has been updated more recently to fix the issue on their end that would mean the `SpecialCustomOperations` item for the `.slnx` file is no longer needed.

CesarD replied:
Ok, now I get it... I would have expected that the .slnx would behave like any other XML file and that they would have prepared the templating engine accordingly... Isn't there any information about this for the templating engine?

matthewtoghill replied:
I've removed the `SpecialCustomOperations` relating to `.slnx` files as it seems it is no longer needed.

matthewtoghill commented on src/Content/Backend/Solution/Directory.Build.props:
```diff
@@ -0,0 +1,9 @@
+    <TargetFramework>net10.0</TargetFramework>
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
+  </PropertyGroup>
```
Can include more properties in here as the defaults for all projects within the solution. If they need to be overridden in a project then we just add those properties back with required values in the relevant `.csproj` files.

CesarD replied:
Yes, this is a very good one I was missing... Since this is an entire solution template, it's good to centralize these things so we don't have all this spread across all the different projects.

CesarD committed:
refactor: Migrate OpenAPI docs from Swagger to Scalar.AspNetCore

CesarD commented:
@matthewtoghill please check the latest changes out... I've replaced the Swashbuckle implementation for Swagger with Scalar and also included the document and operation transformers for the API security so everything can work pretty similar to how it was with Swashbuckle, but without all the custom code for fixing its issues.
Feel free to give it a try and check out the different scenarios. It should exclude the security transformations if Auth flag is disabled.

CesarD committed:
refactor: finished converting the last missing extension methods into block extension methods.

matthewtoghill committed:
chore: update packages

matthewtoghill committed:
chore: remove RequiresAuthentication from integration tests when the auth flag is false as base property is also removed

matthewtoghill committed:
feat: remove sln file

matthewtoghill committed:
chore: match condition to base class

matthewtoghill commented:
> @matthewtoghill please check the latest changes out... I've replaced the Swashbuckle implementation for Swagger with Scalar and also included the document and operation transformers for the API security so everything can work pretty similar to how it was with Swashbuckle, but without all the custom code for fixing its issues. Feel free to give it a try and check out the different scenarios. It should exclude the security transformations if Auth flag is disabled.

Looks good, there was an old issue related to the auth flag being false that is now fixed, or it can be fixed another way if you prefer. When `auth` or `apiService` flags are false then the `IntegrationTest` base class does not include the `RequiresAuthentication` property.

CesarD committed:
Refactor entity ID generation to use app-side GUIDs

EntityTypeConfiguration now uses ValueGeneratedNever for IDs, shifting ID generation from the database to the application (via Guid.NewGuid()). Updated entity constructors, command handlers, and tests to align with this approach. Adjusted EF Core migrations and model snapshots to remove ValueGeneratedOnAdd. Cleaned up related test and factory code for consistency.

matthewtoghill commented on src/Content/Backend/Solution/Monaco.Template.Backend.Common.Api/OpenApi/OAuth2OperationTransformer.cs:
```diff
@@ -0,0 +1,39 @@
+		}
+
+		var requiredScopes = metadata.OfType<IAuthorizeData>()
+									 .Where(a => !string.IsNullOrEmpty(a.Policy))
```
Better to use `string.IsNullOrWhiteSpace`?

CesarD replied:
Done!

CesarD committed:
Refactor Mediator API for CancellationToken & CreatedResponse

Refactored MediatorExtensions to support CancellationToken in all methods, improving async and cancellation handling. Introduced CreatedResponse record for standardized creation responses. Updated Companies, Products, and Files endpoints to use new Mediator methods and return Created<CreatedResponse>. Adjusted integration tests to validate CreatedResponse. Improved code clarity and authorization policy checks.

CesarD committed:
Undo accidental cleanup.

CesarD commented:
@matthewtoghill please review the latest changes I introduced.
Basically, I refactored a little bit the MediatR extensions we had to improve a little bit their naming and also to include the CancellationToken in the async calls being made. I also included a couple more methods for simplifying how to return other types of responses from the endpoint in case of success.

Lastly, please be kind to let me know if you find any issue when running the whole tests battery (including the Integration ones in the same process)... I experienced an issue with the CreateProduct one, where the assertion checking if the ProductCreated message has been consumed is failing, but when I run it individually it succeeds, so I'm not sure if it's my crappy laptop or what, so if you could run it and let me know, that would be great!

Thanks! Cheers!

CesarD committed:
Refactor integration test infra: shared fixture & Respawn

- Use [Collection("IntegrationTests")] to share AppFixture across all test classes, starting containers only once per run
- Add Respawn for fast, reliable DB resets between tests
- Apply DB migrations once in fixture, not per test
- Remove redundant DbContext/migration logic from test base
- Disable test collection parallelization via xunit.runner.schema.json
- Update dependencies and project files for Respawn and config
- Simplify and clarify test setup/teardown for better reliability

CesarD committed:
Update NuGet package versions to latest releases

CesarD committed:
Formatting.

CesarD reviewed (dismissed):
I have added a few fixes to Integration Tests and reconsidered @matthewtoghill 's proposal for implementing Spawner for resetting the DB between test classes. Now Integration Tests will share the same AppFixture so there won't be a constant regeneration of test containers between runs and as a consequence these tests have suffered a huge boost in execution speed.

matthewtoghill committed:
fix: revert change to product test

matthewtoghill committed:
chore: update packages and comment formatting

matthewtoghill committed:
fix: integration test issues

matthewtoghill commented:
> @matthewtoghill please review the latest changes I introduced. Basically, I refactored a little bit the MediatR extensions we had to improve a little bit their naming and also to include the CancellationToken in the async calls being made. I also included a couple more methods for simplifying how to return other types of responses from the endpoint in case of success.
> 
> Lastly, please be kind to let me know if you find any issue when running the whole tests battery (including the Integration ones in the same process)... I experienced an issue with the CreateProduct one, where the assertion checking if the ProductCreated message has been consumed is failing, but when I run it individually it succeeds, so I'm not sure if it's my crappy laptop or what, so if you could run it and let me know, that would be great!
> 
> Thanks! Cheers!

@CesarD - I get the same issue with the test, but this seems to be fixed when I revert the change to lines 334 and 339 from 
`(await apiTestHarness.Published.SelectAsync<ProductCreated>().AnyAsync())`
back to
`(await apiTestHarness.Published.Any<ProductCreated>())`

and

`(await consumerHarness.Consumed.SelectAsync<ProductCreated>().AnyAsync())`
back to
`(await consumerHarness.Consumed.Any<ProductCreated>())`

I've pushed this and some other fixes from my testing. I think we should be there for merging this PR now?

matthewtoghill committed:
chore: remove template config special custom operations for slnx files as no longer needed

CesarD commented:
> > @matthewtoghill please review the latest changes I introduced. Basically, I refactored a little bit the MediatR extensions we had to improve a little bit their naming and also to include the CancellationToken in the async calls being made. I also included a couple more methods for simplifying how to return other types of responses from the endpoint in case of success.
> > 
> > Lastly, please be kind to let me know if you find any issue when running the whole tests battery (including the Integration ones in the same process)... I experienced an issue with the CreateProduct one, where the assertion checking if the ProductCreated message has been consumed is failing, but when I run it individually it succeeds, so I'm not sure if it's my crappy laptop or what, so if you could run it and let me know, that would be great!
> > 
> > Thanks! Cheers!
> 
> @CesarD - I get the same issue with the test, but this seems to be fixed when I revert the change to lines 334 and 339 from 
> `(await apiTestHarness.Published.SelectAsync<ProductCreated>().AnyAsync())`
> back to
> `(await apiTestHarness.Published.Any<ProductCreated>())`
> 
> and
> 
> `(await consumerHarness.Consumed.SelectAsync<ProductCreated>().AnyAsync())`
> back to
> `(await consumerHarness.Consumed.Any<ProductCreated>())`
> 
> I've pushed this and some other fixes from my testing. I think we should be there for merging this PR now?

I actually found out the issue but haven't gotten the chance to finish applying the fix yet:
The Test Harness can only be created and used once per test, if you use it more than once, then it can't assert correctly which was done and where, hence why a test always fails when running the whole test suite and succeeds when running it independently. This is something that I have also already confirmed with the docs and with Chris Patterson himself, so I need to rework the IntegrationTest and AppFixture a little bit in order to organize things in a way that allows to easily generate a new WebAppFactory with the TestHarness registered for the tests that require it. I'll try to have it done this week, if possible.

CesarD committed:
Refactor integration test infra: DI, containers, MassTransit

Refactor and modernize integration test infrastructure:
- Pin Testcontainers images for SQL Server, RabbitMQ, Azurite, Keycloak
- Require IServiceProvider for DbContext resolution; update all usages
- Move MassTransit test harness setup to extension methods
- Add GetCustomFactory for flexible Web/Worker factory config
- Remove direct IHost usage for worker; use DI instead
- Use per-test FlurlClient with proper disposal
- Improve resource cleanup in AppFixture.DisposeAsync
- Use Parallel.ForEachAsync for blob cleanup in ProductsTests
- Enhance MassTransit message assertions for content matching
- General code cleanup and modernization for maintainability and test isolation

CesarD commented:
> > > @matthewtoghill please review the latest changes I introduced. Basically, I refactored a little bit the MediatR extensions we had to improve a little bit their naming and also to include the CancellationToken in the async calls being made. I also included a couple more methods for simplifying how to return other types of responses from the endpoint in case of success.
> > > Lastly, please be kind to let me know if you find any issue when running the whole tests battery (including the Integration ones in the same process)... I experienced an issue with the CreateProduct one, where the assertion checking if the ProductCreated message has been consumed is failing, but when I run it individually it succeeds, so I'm not sure if it's my crappy laptop or what, so if you could run it and let me know, that would be great!
> > > Thanks! Cheers!
> > 
> > 
> > @CesarD - I get the same issue with the test, but this seems to be fixed when I revert the change to lines 334 and 339 from
> > `(await apiTestHarness.Published.SelectAsync<ProductCreated>().AnyAsync())`
> > back to
> > `(await apiTestHarness.Published.Any<ProductCreated>())`
> > and
> > `(await consumerHarness.Consumed.SelectAsync<ProductCreated>().AnyAsync())`
> > back to
> > `(await consumerHarness.Consumed.Any<ProductCreated>())`
> > I've pushed this and some other fixes from my testing. I think we should be there for merging this PR now?
> 
> I actually found out the issue but haven't gotten the chance to finish applying the fix yet: The Test Harness can only be created and used once per test, if you use it more than once, then it can't assert correctly which was done and where, hence why a test always fails when running the whole test suite and succeeds when running it independently. This is something that I have also already confirmed with the docs and with Chris Patterson himself, so I need to rework the IntegrationTest and AppFixture a little bit in order to organize things in a way that allows to easily generate a new WebAppFactory with the TestHarness registered for the tests that require it. I'll try to have it done this week, if possible.

Ok @matthewtoghill, I have finally pushed the latest version with the fixes and refactors needed for the Integration Tests to work correctly. Now they have a bit better isolation and so the instantiation of a new instance of the WebAppFactory and the WorkerServiceFactory allow to register them with the Test Harness only when needed and use it only once so it doesn't get reused across different tests which is what was messing them up in the first place.
Please feel free to test it and let me know if you find anything odd. Integration Tests run super fast now!

BTW: sorry for the long wait, I had to sort some other things out and couldn't spare the time for this, but now it's finally done 😄

CesarD committed:
Resolved merge conflicts.

CesarD committed:
Bump Monaco.Template version to 2.8.0

Updated the version number in Monaco.Template.nuspec from 2.7.0 to 2.8.0 to prepare for the next release. No other changes were made.

CesarD committed:
Exclude .idea directory in template.json patterns

Added .idea to the list of excluded file patterns in template.json to prevent JetBrains IDE configuration files from being included. No other exclusion changes were made.

CesarD committed:
Add optimistic concurrency and 409 conflict support

- Introduce IConcurrencyRetriable marker for safe command retries
- Refactor ConcurrencyExceptionBehavior to handle and retry on DbUpdateConcurrencyException, returning 409 Conflict on failure
- Extend CommandResult/CommandResult<T> with ConcurrencyConflict flag and factory methods
- Update Mediator extensions and API endpoints to support 409 Conflict responses for create, update, and delete operations
- Add rowversion-based Version property to Product entity and update migrations
- Refactor pipeline behavior registration for better coverage
- Update tests and documentation for new concurrency features

CesarD commented:
@matthewtoghill please have a look at my latest commit that I wanted to add a small refactor to the concurrency behavior for being more flexible about having 2 options for dealing with them.
